packages/cluster/tests/substitute-string-keys.js

```diff
@@ -0,0 +1,179 @@
+const test = require('tape')
+
+const { makeDockerImage, substituteVersion } = require('../src/common.js')
+
+const defaults = require('../src/defaults')
+
+const template = defaults.image
+
+test('replace {version} key with a value', function(t) {
+    t.plan(1)
+
+    const string = 'somethingwitha{version}key'
+    const version = '2.31.1'
+
+    const result = substituteVersion(string, version)
+
+    t.equal(result, 'somethingwitha2.31.1key', 'replace {version} with 2.31.1')
+})
+
+test('stable and 2.31.4', function(t) {
+    t.plan(2)
+
+    const expected = 'dhis2/core:2.31.4'
+
+    t.equal(
+        makeDockerImage(template, {
+            version: '2.31.4',
+            channel: 'stable',
+        }),
+        expected,
+        expected
+    )
+
+    t.equal(
+        makeDockerImage(template, {
+            version: '2.31.4',
+        }),
+        expected,
+        expected
+    )
+})
+
+test('stable and 2.32.0 with alpine variant', function(t) {
+    t.plan(1)
+
+    const expected = 'dhis2/core:2.32.0-alpine'
+
+    t.equal(
+        makeDockerImage(
+            template,
+            {
+                version: '2.32.0',
+            },
+            ['alpine']
+        ),
+        expected,
+        expected
+    )
+})
+
+test('stable and master with tomcat9 and debian-slim variant', function(t) {
+    t.plan(1)
+
+    const expected = 'dhis2/core:master-tomcat9-debian-slim'
+
+    t.equal(
+        makeDockerImage(
+            template,
+            {
+                version: 'master',
+            },
+            ['tomcat9', 'debian-slim']
```

Comment:
Question: how does an array like this get passed to the CLI?  Since order is important, how do we enforce that, i.e. linux flavor doesn't get appended before tomcat variant?

Reply:
> Question: how does an array like this get passed to the CLI?

There are three options I can see a user trying, two of which works:

1. `--variant tomcat alpine`
2.`--variant tomcat-alpine` 
3. `--variant tomcat, alpine`

**1** and **2** will result in a correct variant being appended to the image `-tomcat-alpine`. The comma separated option (**3**) does not work.

More info about the `.array()` option: https://github.com/yargs/yargs/blob/master/docs/api.md#arraykey

> Since order is important, how do we enforce that, i.e. linux flavor doesn't get appended before tomcat variant?

Then it is better to force the user to option **2** and take the variant as a single slugged string.

Reply:
Change `variant` to string in aef2f98